File: modulo04-criacao-de-agentes-autonomos-novo/opspilot-java/src/main/java/com/opspilot/team/TeamStrategy.java
package com.opspilot.team;

import com.opspilot.domain.ConversationMessage;
import com.opspilot.domain.ExecutionMetrics;
import com.opspilot.domain.ReasoningStrategy;
import com.opspilot.domain.StrategyRunInput;
import com.opspilot.domain.StrategyResult;

import java.util.EnumMap;
import java.util.Map;

/** Porte de TeamStrategy (team/team-strategy.ts). */
public class TeamStrategy implements ReasoningStrategy {

    private final TeamGraph graph;

    public TeamStrategy(DecideNextFn decideNext, Map<TeamRole, RoleRunner> roleRunners, int supervisorLlmCalls) {
        this.graph = new TeamGraph(decideNext, new EnumMap<>(roleRunners), supervisorLlmCalls);
    }

    @Override
    public String name() {
        return "team";
    }

    @Override
    public StrategyResult run(StrategyRunInput input) {
        long startedAt = System.currentTimeMillis();
        TeamGraph.Result result = graph.run(composeMessage(input));
        return new StrategyResult(
                result.answer(),
                result.trace(),
                new ExecutionMetrics(result.llmCalls(), System.currentTimeMillis() - startedAt));
    }

    private static String composeMessage(StrategyRunInput input) {
        if (input.history().isEmpty()) {
            return input.message();
        }
        StringBuilder sb = new StringBuilder("Historico recente da conversa:\n");
        for (ConversationMessage message : input.history()) {
            sb.append(message.role()).append(": ").append(message.content()).append('\n');
        }
        sb.append('\n').append(input.message());
        return sb.toString();
    }
}

File: modulo04-criacao-de-agentes-autonomos-novo/opspilot-go/team/team_strategy.go
package team

import (
	"strings"
	"time"

	"opspilot/domain"
)

// TeamStrategy é o porte de TeamStrategy (team/team-strategy.ts): expõe o
// TeamGraph como um domain.ReasoningStrategy, ao lado de strategies.ReactStrategy.
type TeamStrategy struct {
	graph *TeamGraph
}

var _ domain.ReasoningStrategy = (*TeamStrategy)(nil)

func NewTeamStrategy(decideNext DecideNextFn, roleRunners map[Role]RoleRunner, supervisorLLMCalls int) *TeamStrategy {
	return &TeamStrategy{graph: NewTeamGraph(decideNext, roleRunners, supervisorLLMCalls)}
}

func (s *TeamStrategy) Name() string { return "team" }

func (s *TeamStrategy) Run(input domain.StrategyRunInput) (domain.StrategyResult, error) {
	startedAt := time.Now()
	result := s.graph.Run(composeMessage(input))
	return domain.StrategyResult{
		Answer: result.Answer,
		Trace:  result.Trace,
		Metrics: domain.ExecutionMetrics{
			LLMCalls:  result.LLMCalls,
			LatencyMs: time.Since(startedAt).Milliseconds(),
		},
	}, nil
}

func composeMessage(input domain.StrategyRunInput) string {
	if len(input.History) == 0 {
		return input.Message
	}
	var sb strings.Builder
	sb.WriteString("Historico recente da conversa:\n")
	for _, message := range input.History {
		sb.WriteString(message.Role)
		sb.WriteString(": ")
		sb.WriteString(message.Content)
		sb.WriteString("\n")
	}
	sb.WriteString("\n")
	sb.WriteString(input.Message)
	return sb.String()
}
